### ISSUE TITLE
Gradle instance currently in use. HPC

### ISSUE THREAD

**Issue by aadharna** [gradle/gradle#12231]
tl;dr: I wrote some `python` code to take advantage of an High Performance Computer (HPC) by writing in a parent-child paradigm. Each child creates a single java simulator and the parent creates a bunch of java simulators (which all uses Gradle) and launching those simulators is having weird behaviour on an HPC.

The HPC has hundreds of CPUs and GB of RAM. However, there is one underlying file-system. 

One of the admins on the HPC told me: 
> On HPC cluster, please be careful with file lock, it is not reliable sometimes. For example, file lock on /scratch only works for processes running on same compute node, file lock will not work across compute node. Even for /home, file lock on one compute node will not immediately available from other nodes, network file systems.

----  

I have created a stack-overflow post here with details: https://stackoverflow.com/questions/60198505/gradle-instance-currently-in-use-hpc 

When launching parent-child worker scripts on an HPC where the parent can hold n simulators (which are written in java) and where n children each have 1 simulator, even though the children are on separate processes, it seems like they are conflicting over a .lock file. Specifically: `/home/ad5238/.gradle/caches/5.1.1/fileHashes/fileHashes.lock` 

The behavior persists even upon a complete `rm -rf` clear of the cache files and a rebuild of gradle through `./gradlew clean build --parallel`.

### Expected Behavior

On a "regular" machine, when I launch my scripts, my parent script will create (potentially many) Java simulators. The parent will then dole out work to each available child who will: read in that work, update its simulator to the specifications sent by the parent, execute the simulation, and send back information. 

If parent has four tasks and three children are alive, the parent will send two tasks to one child, and the rest will get a single task.

### Current Behavior

It seems that upon launch of a simulator in the HPC, the simulators are fighting over a specific .lock file. (See attached failure.log file). This causes the build phase of the simulators to fail and therefore the scripts do not launch. 

### Context

This issue renders my code completely unable to execute on an HPC. This is a problem because although it works on a "normal" computer, normal computers don't have the CPU capabilities necessary to speed up my work such that I will get meaningful results in time for my thesis. 

### Steps to Reproduce 

Unfortunately this is not reproducible without an HPC as on normal computers I am not running into this problem. 

That being said, you can see my github project here: https://github.com/aadharna/UntouchableThunder 

In MY code, the launching of the java simulators can be found here: https://github.com/aadharna/UntouchableThunder/blob/df4e3ecd2f8ae4b9a30d2f53ae9af06c95c79a04/generator/env_gen_wrapper.py#L64

The actual launching of the java-simulators can be found here: https://github.com/aadharna/GVGAI_GYM/blob/05c0447c1da48e713375819c84c2be8bac2fb799/python/gvgai/gvgai/client/gvgai_client.py#L37

The relevant `python` files from my code are: `generator/env_gen_wrapper.py.` The following files implement the communication between parent and child. They shouldn't be neccessary, but just in case these are where they are. The first two are the ones that get launched from command-line. If you were to look at one of these, I would suggest `child.py` and the paired `util` file. [`poet_distributed.py`, `child.py`, `utils/ADPParent.py` and `utils/ADPChild.py`]

----

Running on the HPC is invoked by running `bash runone.sh [EXPERIMENT_NAME]`

On a normal computer you would launch each command from a new terminal of: 
```
python poet_distributed.py --exp_name ??
python child.py --id 1
python child.py --id 2
...
python child.py --id n
```

----  

The library which i am using which overwrites the normal OpenAI gym constructor is here: https://github.com/aadharna/GVGAI_GYM/tree/05c0447c1da48e713375819c84c2be8bac2fb799 

### Your Environment

Build scan URL: https://gradle.com/s/4pqarpxugadso

This is on an HPC. You can find details about it here: https://devwikis.nyu.edu/display/NYUHPC/High+Performance+Computing+at+NYU

---- 

[failure.log](https://github.com/gradle/gradle/files/4201992/failure.log)

**Comment by big-guy** [comment]
IIUC, I think you're running into problems because the different Gradle processes are running on different nodes but share a common filesystem where the Gradle User Home is.  The Gradle process uses IPC to communicate to other daemons on the same machine to coordinate lock accesses and share common directories.  This only works on the same machine where local communication is possible.

I think it may be possible to make this work by having each Gradle process use a separate Gradle User Home (You can change this with `-g`).   I can't quite tell how everything is wired together, but if the Gradle build isn't building something different each time, you could use Gradle to build a distribution first and then run the java-simulator with the script generated by the Gradle build.  That wouldn't have any locking problems.

Good luck on the thesis!

**Comment by aadharna** [comment]
First off, thanks for the response!

Would I add the `-g` flag here: https://github.com/aadharna/GVGAI_GYM/blob/05c0447c1da48e713375819c84c2be8bac2fb799/python/gvgai/gvgai/client/gvgai_client.py#L37

Two things:   
- I've reworked the code such that my "main/master" loop doesn't actually spawn any java(gradle) simulators and only the children do. In case that's helpful information. 
- One of the admin's on the HPC has had me set up a Singularity image. When inside that image, my scripts managed to run (it was only a test, but it proves the main code can run)! Why do you think that running inside the Singularity image makes a difference, and would it still be worth me making this change you've described?

**Comment by big-guy** [comment]
Right, I think you could change the arguments to gradlew here:
https://github.com/aadharna/GVGAI_GYM/blob/05c0447c1da48e713375819c84c2be8bac2fb799/python/gvgai/gvgai/client/gvgai_client.py#L44-L45

You'd specify it like `gradlew -g /path/to/another/directory` where `/path/to/another/directory` would need to be unique for each different node.

> I've reworked the code such that my "main/master" loop doesn't actually spawn any java(gradle) simulators and only the children do. In case that's helpful information.

It looks to me that you're just running `gradlew run` on the clients with different arguments being passed to the underlying application (not Gradle, but the Java application you built with Gradle).  If that's the case, you don't need to run Gradle every time.  IIUC, it looks like it's possible to have the "master" run Gradle once (e.g., run `gradlew install`) and since all of the client nodes share the same filesystem, you can just run the Java application directly without going through Gradle. 

After running `gradlew install`, you should have a script under the project directory: `build//install/<name of your project>/bin/<name of your project>`.  e.g., if the name of your project is `java-app`, then you can run `build//install/java-app/bin/java-app` (it's just a shell script).

**Comment by aadharna** [comment]
I created a directory inside the `.gradle` folder to house each unique gradle-home. It is normal for the gradle-homes to be empty? 
![image](https://user-images.githubusercontent.com/16660757/74613869-a89d5100-50e0-11ea-9829-91ba3acadd3d.png)

**Comment by aadharna** [comment]
This has been closed with the `./gradlew install` solution having been implemented.
